src/librustc_typeck/collect.rs

```diff
@@ -2898,13 +2905,15 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
                 codegen_fn_attrs.export_name = Some(s);
             }
         } else if attr.check_name(sym::target_feature) {
-            if tcx.fn_sig(id).unsafety() == Unsafety::Normal {
-                let msg = "`#[target_feature(..)]` can only be applied to `unsafe` functions";
-                tcx.sess
-                    .struct_span_err(attr.span, msg)
-                    .span_label(attr.span, "can only be applied to `unsafe` functions")
-                    .span_label(tcx.def_span(id), "not an `unsafe` function")
-                    .emit();
+            if !tcx.is_closure(id) {
+                if tcx.fn_sig(id).unsafety() == Unsafety::Normal {
```

Comment:
Closures are not `unsafe`, so the error should be reported for them as well.

Reply:
Addressed as well.